Public/Src/Pips/Dll/Operations/PipGraphFragmentSerializer.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using BuildXL.Utilities;
+
+namespace BuildXL.Pips.Operations
+{
+    /// <summary>
+    /// Class to handle serialization/deserialization of pip graph fragments
+    /// </summary>
+    public class PipGraphFragmentSerializer
+    {
+        /// <summary>
+        /// Total pips in the fragment.  Until this is read in from the file, it is 0.
+        /// </summary>
+        public int TotalPips { get; private set; }
+
+        /// <summary>
+        /// Total pips deserialized so far.
+        /// </summary>
+        public int PipsDeserialized { get; private set; }
+
+        /// <summary>
+        /// Total pips serialized so far.
+        /// </summary>
+        public int PipsSerialized { get; private set; }
+
+        /// <summary>
+        /// Description of the fragment, for printing on the console
+        /// </summary>
+        public string FragmentDescription { get; private set; }
+
+        /// <summary>
+        /// Deserialize a pip graph fragment and call the given handleDeserializedPip function on each pip deserialized
+        /// Returns true if successfully handled all pips.
+        /// </summary>
+        public bool Deserialize(string fragmentDescription, PipExecutionContext context, PipGraphFragmentContext pipFragmentContext, AbsolutePath filePath, Func<Pip, bool> handleDeserializedPip)
+        {
+            FragmentDescription = fragmentDescription;
+            PipsDeserialized = 0;
+            string fileName = filePath.ToString(context.PathTable);
+            using (FileStream stream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (RemapReader reader = new RemapReader(pipFragmentContext, stream, context))
+            {
+                while (reader.ReadBoolean())
+                {
+                    var pip = Pip.Deserialize(reader);
+                    if (handleDeserializedPip != null)
+                    {
+                        if (!handleDeserializedPip.Invoke(pip))
+                        {
+                            return false;
+                        }
+                    }
+
+                    PipsDeserialized++;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Serialize list of pips to a file
+        /// </summary>
+        public void Serialize(string fragmentName, PipExecutionContext context, AbsolutePath filePath, IEnumerable<Pip> pipsToSerialize)

```

Comment:
>IEnumerable<Pip> pipsToSerialize [](start = 103, length = 32)

Does this need to be IEnumerable? If you made it a list or an array you could serialize the count rather than the boolean for each item to optimize the serialization a bit

Reply:
Yeah, I intend to write the count.  Currently I don't have a tool to write the fragments anymore (I removed it for the PR because it required making some internal things public).   So I am testing with existing fragments and using that format.

---
In reply to: [289586823](https://github.com/microsoft/BuildXL/pull/394#discussion_r289586823) [](ancestors = 289586823)